README.md
# Platty-Game
A short platformer game that i made for a school project
I made this game as a challenge for myself to make a working game in 48 hours...

it's not amazing but for the very first project that i did by myself im happy with it :))

to download the game you can clone the whole repo and play the platfformer game.exe..

no python or anything else needed


Changelog.txt
DIBUAT UNTUK BERSENANG SENANG>>

to play the game run the PlatformerGame.api


sprites.py
# tempat character dan semua sprite untuk JUMPS!
import pygame as pg
from constants import *
from random import choice, randrange
vct = pg.math.Vector2

## Class untuk mengambil spritesheet #####
class SpriteSheet(pg.sprite.Sprite):
    def __init__(self, filename):
        self.spritesheet = pg.image.load(filename).convert()

    def get_image(self, x, y, width, height):
        # fungsi untuk mengambil image dari spritesheet

        image = pg.Surface((width, height))
        image.blit(self.spritesheet, (0,0), (x , y, width, height))
        image = pg.transform.scale(image, (int(width * 0.75), int(height * 0.75)))
        return image

    def get_image_half(self, x, y, width, height):
        # mengambil image dari spritesheet dengan setengah scaling

        image = pg.Surface((width, height))
        image.blit(self.spritesheet, (0,0), (x , y, width, height))
        image = pg.transform.scale(image, (width // 2, height // 2))
        return image

## Class player untuk pemain #############
class Player(pg.sprite.Sprite):
    def __init__(self, game):
        self._layer = PLAYER_LYR
        self.groups = game.allSprite
        pg.sprite.Sprite.__init__(self, self.groups)
        self.game = game
        self.walking = False
        self.jumping = False
        self.boost_pwr = False
        self.current_fr = 0
        self.last_upd = 0
        self.load_img()
        self.image = self.standing_frame[0]
        self.rect = self.image.get_rect()
        self.rect.center = (40, HEIGHT - 100)
        self.pos = vct(40, HEIGHT - 100)
        self.vel = vct(0, 0)
        self.acc = vct(0, 0)

    def load_img(self):
        # image dari player

        self.standing_frame = [self.game.spritesheet_player1.get_image(0, 190, 66, 92),
                                self.game.spritesheet_player1.get_image(67, 190, 66, 92)]
        for fr in self.standing_frame:
            fr.set_colorkey(BLACK)
        
        self.walking_r = [self.game.spritesheet_player1.get_image(142, 0, 70, 94),
                            self.game.spritesheet_player1.get_image(0, 95, 70, 94),
                            self.game.spritesheet_player1.get_image(71, 95, 70, 94),
                            self.game.spritesheet_player1.get_image(142, 95, 70, 94),
                            self.game.spritesheet_player1.get_image(213, 0, 70, 94),
                            self.game.spritesheet_player1.get_image(284, 0, 70, 94)]
        
        self.walking_l = []
        for fr in self.walking_r:
            fr.set_colorkey(BLACK)
            self.walking_l.append(pg.transform.flip(fr, True, False))

        self.jumping_frame_r = self.game.spritesheet_player1.get_image(423, 95, 66, 94)
        self.jumping_frame_r.set_colorkey(BLACK)    
        self.jumping_frame_l = pg.transform.flip(self.jumping_frame_r, True, False)
        self.jumping_frame_l.set_colorkey(BLACK)    
    
    def jump(self):
        # fungsi untuk melompat

        self.rect.x += 2
        hit = pg.sprite.spritecollide(self, self.game.platforms, False)
        self.rect.x -= 2
        if hit and not self.jumping:
            self.game.jump_snd.play()
            self.jumping = True
            self.vel.y = -PLAYER_JMP

    def jump_stop(self):
        # fungsi untuk menstop lompatan

        if self.jumping:
            if self.vel.y < -3:
                self.vel.y = -3

    def update(self):
        # fungsi update dari player

        self.animasi()
        self.acc = vct(0, PLAYER_GRV)
        key = pg.key.get_pressed()
        if key[pg.K_LEFT]:
            self.acc.x = -PLAYER_ACC
        if key[pg.K_RIGHT]:
            self.acc.x = PLAYER_ACC

        # menambah friksi ke player
        self.acc.x += self.vel.x * PLAYER_FRC

        # equasi motion
        self.vel += self.acc
        if abs(self.vel.x) < 0.1:
            self.vel.x = 0
        self.pos += self.vel + 0.5 * self.acc

        # agar ga keluar frame screen dari game
        if self.pos.x > WIDTH + self.rect.width / 2:
            self.pos.x = 0 - self.rect.width / 2
        if self.pos.x < 0 - self.rect.width / 2:
            self.pos.x = WIDTH + self.rect.width / 2

        self.rect.midbottom = self.pos


    def animasi(self):
        # fungsi untuk membuat animasi dari player

        now = pg.time.get_ticks()
        if self.vel.x != 0:
            self.walking = True
        else:
            self.walking = False

        # animasi pas jalan
        if self.walking:
            if now - self.last_upd > 100:
                self.last_upd = now
                self.current_fr = (self.current_fr + 1) % len(self.walking_l)
                bot = self.rect.bottom
                if self.vel.x > 0:
                    self.image = self.walking_r[self.current_fr]
                else:
                    self.image = self.walking_l[self.current_fr]
                self.rect = self.image.get_rect()
                self.rect.bottom = bot
                

        # animasi pas idle(diam)
        if not self.jumping and not self.walking:
            if now - self.last_upd > 345:
                self.last_upd = now
                self.current_fr = (self.current_fr + 1) % len(self.standing_frame)
                bot = self.rect.bottom
                self.image = self.standing_frame[self.current_fr]
                self.rect = self.image.get_rect()
                self.rect.bottom = bot

        # animasi pas lompat
        if self.jumping or self.boost_pwr:
            if now - self.last_upd > 50:
                self.last_upd = now
                bot = self.rect.bottom
                if self.vel.x > 0:
                    self.image = self.jumping_frame_r
                else:
                    self.image = self.jumping_frame_l

                self.rect = self.image.get_rect()
                self.rect.bottom = bot

        # membuat mask untuk collision
        self.mask = pg.mask.from_surface(self.image)

## class untuk platform di game ##########
class Platform(pg.sprite.Sprite):
    def __init__(self, game, x, y,):
        self._layer = PLAT_LYR
        self.groups = game.platforms, game.allSprite
        pg.sprite.Sprite.__init__(self, self.groups)
        self.game = game
        images = [self.game.spritesheet_objects.get_image_half(0, 288, 380, 94),
                    self.game.spritesheet_objects.get_image_half(213, 1662, 201, 100),
                    self.game.spritesheet_objects.get_image_half(0, 384, 380, 94),
                    self.game.spritesheet_objects.get_image_half(382, 204, 200, 100)]
        images_2 = [self.game.spritesheet_objects.get_image_half(0, 960, 380, 94),
                    self.game.spritesheet_objects.get_image_half(0, 864, 380, 94),
                    self.game.spritesheet_objects.get_image_half(218, 1558, 200, 100),
                    self.game.spritesheet_objects.get_image_half(382, 0, 200, 100)]
        if randrange(100) < PLATFORM_SPWN:
            self.image = choice(images_2)
        else:
            self.image = choice(images)
        self.image.set_colorkey(BLACK)
        self.rect = self.image.get_rect()
        self.rect.y = y
        self.rect.x = x

        # untuk men-spawn powerup
        if randrange(100) < PWR_SPAWN:
            Pwup(self.game, self)

## Class untuk powerup di game ###########
class Pwup(pg.sprite.Sprite):
    def __init__(self, game, plat):
        self._layer = PWUP_LYR
        self.groups = game.powerup, game.allSprite
        pg.sprite.Sprite.__init__(self, self.groups)
        self.game = game
        self.plat = plat
        self.type = choice(['boost'])
        self.image = self.game.spritesheet_objects.get_image_half(820, 1805, 71, 70)
        self.image.set_colorkey(BLACK)
        self.rect = self.image.get_rect()
        self.rect.centerx = self.plat.rect.centerx
        self.rect.bottom = self.plat.rect.top - 5

    def update(self):
        # fungsi update untuk powerup

        self.rect.bottom = self.plat.rect.top - 5
        if not self.game.platforms.has(self.plat):
            self.kill()


## Class untuk Enemy di game #############
class Enemy(pg.sprite.Sprite):
    def __init__(self, game):
        self._layer = ENEMIES_LYR        
        self.groups = game.allSprite, game.Enemy
        pg.sprite.Sprite.__init__(self, self.groups)
        self.game = game
        self.image_up_l = self.game.spritesheet_enemies.get_image(0, 32, 72, 36)
        self.image_up_l.set_colorkey(BLACK)
        self.image_down_l = self.game.spritesheet_enemies.get_image(0, 0, 75, 31)
        self.image_down_l.set_colorkey(BLACK)
        self.image_up_r = pg.transform.flip(self.image_up_l, True, False)
        self.image_up_r.set_colorkey(BLACK)        
        self.image_down_r = pg.transform.flip(self.image_down_l, True, False)
        self.image_down_r.set_colorkey(BLACK)


        self.image = self.image_up_l
        self.rect = self.image.get_rect()
        self.rect.centerx = choice([-100, WIDTH + 100])
        self.vx = randrange(1, 4)
        if self.rect.centerx > WIDTH:
            self.vx *= -1
        self.rect.y = randrange(HEIGHT / 2)
        self.vy = 0
        self.dy = 0.5

    def update(self):
        # fungsi update buat enemy

        self.rect.x += self.vx
        self.vy += self.dy
        if self.vy > 3 or self.vy < -3:
            self.dy *= -1
        center = self.rect.center
        if self.vx < 0:
            if self.dy < 0:
                self.image = self.image_up_l
            else:
                self.image = self.image_down_l
        else:
            if self.dy < 0:
                self.image = self.image_up_r
            else:
                self.image = self.image_down_r
        self.rect = self.image.get_rect()
        self.mask = pg.mask.from_surface(self.image)
        self.rect.center = center
        self.rect.y += self.vy

        # untuk menghapus enemy jika melewati screen
        if self.rect.left > WIDTH + 100 or self.rect.right < -100:
            self.kill()

## Class untuk awan di game background ###
class Cloud(pg.sprite.Sprite):
    def __init__(self, game):
        self._layer = CLOUD_LYR
        self.groups = game.allSprite, game.bg
        pg.sprite.Sprite.__init__(self, self.groups)
        self.game = game
        self.image = choice(self.game.cloud_img)
        self.image.set_colorkey(BLACK)
        self.rect = self.image.get_rect()
        scale = randrange(50, 101) / 100
        self.image = pg.transform.scale(self.image, (int(self.rect.width * scale), int(self.rect.height  * scale)))
        self.rect.x = randrange(WIDTH - self.rect.width)
        self.rect.y = randrange(-500, -30)

    def update(self):
        # fungsi update dari awan di game

        # untuk menghapus awan jika melewati batas screen tertentu
        if self.rect.top > HEIGHT * 2:
            self.kill()

constants.py
# game option/settings and stuff
TITLE = 'Jumps!'
WIDTH = 480
HEIGHT = 600
FPS = 60
FONT_NAME = 'arial'
HIGHSCORE = 'Hs.txt'
SPRITESHEET = ['p2_spritesheet.png',
                'spritesheet_jumper.png',
                'enemies_spritesheet.png']

# Player properties
PLAYER_ACC = 0.5
PLAYER_FRC = -0.12
PLAYER_GRV = 0.8
PLAYER_JMP = 20


# platform
PLATFORM_LST = [
        (0, HEIGHT - 60),
        (WIDTH / 2 - 50, HEIGHT * 3/4),
        (125, HEIGHT - 350),
        (350, 200),
        (175, 100)
    ]

# game properties
BOOST_PWR = 60
PWR_SPAWN = 5
PLATFORM_SPWN = 20
ENEMY_SPWN = 5000
PLAYER_LYR = 2
PWUP_LYR = 1
PLAT_LYR = 1
ENEMIES_LYR = 2
CLOUD_LYR = 0


# define colors
BLACK = (0, 0, 0) 
WHITE = (255, 255, 255) 
RED = (255, 0, 0) 
GREEN = (0, 255, 0) 
BLUE = (0, 0, 255) 
YELLOW = (255, 255, 0)
LIGHTBLUE = (0, 155, 155)
SKYBLUE = (135, 206, 250)
DEEPSKYBLUE = (0, 19, 255)
DEEPSKYBLUE1 = (0, 75, 100)
ANOTHERBLUE = (173, 216, 230)

BG = DEEPSKYBLUE1

main.py
# JUMPS! by mushaffa huda - platform game
# Art by Kenney.nl
# Happy by https://opengameart.org/users/rezoner
# Happy Adventure by https://opengameart.org/users/TinyWorlds

import pygame as pg
import random
from constants import *
from sprites import *
from os import path

##### ENDLESS PLATFROM GAME #######

class Game:

    def __init__(self):
        pg.init()
        pg.mixer.init()
        pg.display.set_caption(TITLE)
        self.screen = pg.display.set_mode((WIDTH, HEIGHT))
        self.clock = pg.time.Clock()
        self.gamerun = True
        self.font_name = pg.font.match_font(FONT_NAME)
        self.load()

    def load(self):
        # load data from the system to the game

        # load highscore
        self.dir = path.dirname(__file__)
        img_dir = path.join(self.dir, 'img')
        with open(path.join(self.dir, HIGHSCORE), 'w') as r:
            try:
                self.highscore = int(r.read())
            except: 
                self.highscore = 0

        # load sprite image
        self.spritesheet_player1 = SpriteSheet(path.join(img_dir, SPRITESHEET[0]))
        self.spritesheet_objects = SpriteSheet(path.join(img_dir, SPRITESHEET[1]))
        self.spritesheet_enemies = SpriteSheet(path.join(img_dir, SPRITESHEET[2]))

        # load sound files
        self.snd_dir = path.join(self.dir, 'snd')
        self.jump_snd = pg.mixer.Sound(path.join(self.snd_dir, 'Jump5.wav'))
        self.boost_pwup_snd = pg.mixer.Sound(path.join(self.snd_dir, 'boost.wav'))
        self.fall_snd = pg.mixer.Sound(path.join(self.snd_dir, 'Hit_Hurt3.wav'))
        self.hit_snd = pg.mixer.Sound(path.join(self.snd_dir, 'Hit_Hurt.wav'))

        # load clouds
        self.cloud_img = []
        for i in range(1, 4):
            self.cloud_img.append(pg.image.load(path.join(img_dir, 'cloud{}.png'.format(i))).convert())

    def new_game(self):
        # initiate new game

        self.points = 0
        self.allSprite = pg.sprite.LayeredUpdates()
        self.platforms = pg.sprite.Group()
        self.powerup = pg.sprite.Group()
        self.Enemy = pg.sprite.Group()
        self.bg = pg.sprite.Group()
        self.player = Player(self)
        self.enemy_timer = 0
        for pl in PLATFORM_LST:
            Platform(self, *pl)
        pg.mixer.music.load(path.join(self.snd_dir, 'happy.ogg'))
        for i in range(7):
            c = Cloud(self)
            c.rect.y += 450
        self.run_game()

    def run_game(self):
        # ngeloop gamenya

        pg.mixer.music.play(loops = -1)
        self.play = True
        while self.play:
            self.clock.tick(FPS)
            self.update()
            self.events()
            self.draw()
        pg.mixer.music.fadeout(500)

    def update(self):
        # update part dari gamenya

        self.allSprite.update()

        # player stand on platform only if falling
        if self.player.vel.y > 0:
            hit = pg.sprite.spritecollide(self.player, self.platforms, False)
            if hit:
                lowest = hit[0]
                for i in hit:
                    if i.rect.bottom > lowest.rect.bottom:
                        lowest = i
                if lowest.rect.left - 10 < self.player.pos.x < lowest.rect.right + 10:
                    if self.player.pos.y < lowest.rect.centery:
                        self.player.pos.y = hit[0].rect.top + 1
                        self.player.vel.y = 0
                        self.player.jumping = False
                        self.player.boost_pwr = False
        
        # player reach 1/4 top of screen
        if self.player.rect.top <= HEIGHT / 4:
            if random.randrange(100) < 10:
                Cloud(self)
            self.player.pos.y += max(abs(self.player.vel.y), 2)
            for c in self.bg:
                c_velo = random.randrange(2, 5)
                c.rect.y += max(abs(self.player.vel.y / c_velo), 2)
            for e in self.Enemy:
                e.rect.y += max(abs(self.player.vel.y), 2)
            for p in self.platforms:
                p.rect.y += max(abs(self.player.vel.y), 2)
                if p.rect.top >= HEIGHT:
                    p.kill()
                    self.points += 10

        # spawning new platforms
        while len(self.platforms) < 6:
            width = random.randrange(45, 150)
            Platform(self, random.randrange(0, WIDTH - width),
                        random.randrange(-50, -30),
                    )

        # spawn enemy
        now = pg.time.get_ticks()
        if now - self.enemy_timer > 5000 + random.choice([-1000, -500, 0, 500, 1000]):
            self.enemy_timer = now
            Enemy(self)

        # player hit powerup
        Pwup_hit = pg.sprite.spritecollide(self.player, self.powerup, True)
        for p in Pwup_hit:
            if p.type == 'boost':
                self.boost_pwup_snd.play()
                self.player.vel.y = -BOOST_PWR
                self.player.jumping = False
                self.player.boost_pwr = True

        # player hit enemy
        enem_hit = pg.sprite.spritecollide(self.player, self.Enemy, False, pg.sprite.collide_mask)
        if enem_hit:
            self.hit_snd.play()
            self.play = False

        # GAMEOVER!
        if self.player.rect.bottom > HEIGHT:
            for i in self.allSprite:
                i.rect.y -= max(self.player.vel.y, 10)
                if i.rect.bottom < 0:
                    i.kill()
            if len(self.platforms) == 0:
                self.fall_snd.play()
                self.play = False

    def events(self):
        # game events - player input

        for e in pg.event.get():
            # check for closing window using GUI
            if e.type == pg.QUIT:
                if self.play:
                    self.play = False
                self.gamerun = False
            if e.type == pg.KEYDOWN:
                if e.key == pg.K_SPACE or e.key == pg.K_UP:
                    self.player.jump()
            if e.type == pg.KEYUP:
                if e.key == pg.K_SPACE or e.key == pg.K_UP:
                    self.player.jump_stop()
    def draw(self):
        # drawing the game in the screen

        self.screen.fill(BG)
        self.allSprite.draw(self.screen)
        self.drawtext(str(self.points), 22, WHITE, WIDTH / 2, 15)
        # do this last
        pg.display.flip()

    def start_screen(self):
        # start menu screen

        pg.mixer.music.load(path.join(self.snd_dir, 'happy_adventure.ogg'))
        pg.mixer.music.play(loops = -1)
        self.screen.fill(BG)
        self.drawtext(TITLE, 56, WHITE, WIDTH / 2, HEIGHT / 4)
        self.drawtext('Use Arrow key to move and spacebar to jump', 22, WHITE, WIDTH / 2, HEIGHT / 2)
        self.drawtext('Press any key to play', 22, WHITE, WIDTH / 2, HEIGHT * 3 / 4)
        self.drawtext('High Score: ' + str(self.highscore), 22, WHITE, WIDTH / 2, 15)
        self.drawtext('credits: ', 15, WHITE, 25, HEIGHT - 60)
        self.drawtext('-Art by Kenney.nl', 15, WHITE, 50, HEIGHT - 40)
        self.drawtext('-Music by rezoner and TinyWorlds', 15, WHITE, 99, HEIGHT - 20)
        self.drawtext('-by Mushaffa Huda', 15, WHITE, WIDTH / 2, 210)
        pg.display.flip()
        self.waitKey_event()
        pg.mixer.music.fadeout(500)

    def go_screen(self):
        # game over screen

        if self.gamerun:
            pg.mixer.music.load(path.join(self.snd_dir, 'happy_adventure.ogg'))
            pg.mixer.music.play(loops = -1)
            self.screen.fill(BG)
            self.drawtext('GAME OVER CUY!', 48, WHITE, WIDTH / 2, HEIGHT / 4)
            self.drawtext('Score : ' + str(self.points), 22, WHITE, WIDTH / 2, HEIGHT / 2)
            self.drawtext('Press any key to play again', 22, WHITE, WIDTH / 2, HEIGHT * 3 / 4)
            if self.points > self.highscore:
                self.highscore = self.points
                self.drawtext('NEW HIGHSCORE MANTAP!', 22, WHITE, WIDTH / 2, HEIGHT / 2 + 40) 
                with open(path.join(self.dir, HIGHSCORE), 'w') as r:
                    r.write(str(self.points))
            else:
                self.drawtext('High Score: ' + str(self.highscore), 22, WHITE, WIDTH / 2, HEIGHT / 2 + 40)
            pg.display.flip()
            self.waitKey_event()
            pg.mixer.music.fadeout(500)
        return

    def waitKey_event(self):
        # menunggu pemain untuk memencet key di start atau gameover screen

        wait = True
        while wait:
            self.clock.tick(FPS)
            for e in pg.event.get():
                if e.type == pg.QUIT:
                    wait = False
                    self.gamerun = False
                if e.type == pg.KEYUP:
                    wait = False
                    
    def drawtext(self, text, size, color, x, y):
        # fungsi untuk menggambar text ke screen

        font = pg.font.Font(self.font_name, size)
        text_srf = font.render(text, True, color)
        text_rect = text_srf.get_rect()
        text_rect.midtop = (x, y)
        self.screen.blit(text_srf, text_rect)

game = Game()
game.start_screen()

# the game loop to run the game
while game.gamerun:
    game.new_game()
    game.go_screen()

pg.quit()   # quitting the game :)